PFW-1504 Add visualization on LCD

gudnimg opened the pull request:
This PR adds visualization for the try-load-unload sequence.

Change in memory:
Flash: +194 bytes
SRAM: 0 bytes

gudnimg committed:
PFW-1504 initial commit

Change in memory:
Flash: +162 bytes
SRAM: 0 bytes

gudnimg committed:
PFW-1504 Update last position as quickly as possible

0.1mm extrusion may happen while updating the LCD

No change in memory

gudnimg committed:
PFW-1504 Add init for progress bar

In case we are running a retry, the firmware
should clear the old rendering before
starting on a new one

Change in memory:
Flash: +6 bytes
SRAM: 0 bytes

gudnimg committed:
PFW-1504 too much?

I'm not happy with the current solution, I think we can improve it
by using the fact that this is a triangle wave. Need to think about it.

gudnimg committed:
fixup

DRracer commented on Firmware/mmu2_marlin1.cpp:
```diff
@@ -27,7 +27,7 @@ bool planner_any_moves() {
 
 float planner_get_machine_position_E_mm(){
-    return current_position[E_AXIS];
+    return st_get_position_mm(E_AXIS);
```
we should introduce another function `stepper_get_machine_position_E_mm`

gudnimg replied:
Fixed in [9f02bec](https://github.com/prusa3d/Prusa-Firmware/pull/4076/commits/9f02bec7f293092cd5acd811cbe01e622e1f2534)

DRracer commented on Firmware/mmu2_reporting.cpp:
```diff
@@ -289,6 +289,20 @@ void ReportProgressHook(CommandInProgress cip, uint16_t ec) {
+    lcd_space(LCD_WIDTH);
+
+    // Reset cursor position
+    lcd_set_cursor(0, 3);
```
Are you sure no other drawing function accesses the LCD while doing the progress bar?
Other functions may screw the cursor and we may start printing to some other position.

gudnimg replied:
That's a good point. I'll add back `lcd_putc_at`, it will increase code size probably but it should be safer for now.

gudnimg replied:
Fixed with [78ea5ee](https://github.com/prusa3d/Prusa-Firmware/pull/4076/commits/78ea5ee5da5b330877e0c17beff5c52b0df77a68)

gudnimg committed:
PFW-1504 implement stepper_get_machine_position_E_mm()

This function should be used when reading the current machine
position while a move is ongoing

No change in memory

gudnimg committed:
PFW-1504 Set cursor position each time to be safe

Change in memory:
Flash: +18 bytes
SRAM: 0 bytes

gudnimg committed:
Fixup

We're always setting the cursor position when rendering now

Saves 8 bytes of flash

DRracer reviewed (approved):
looks good to me :+1:

gudnimg commented:
There is one issue with my current implementation. Sometimes it will not render the last pixel due to delays in the while loop. I'm currently looking into how to correct this behavior.

gudnimg committed:
PFW-1504 Add precision error

This fixes an issue where sometimes
not all 20 pixels are rendered.

It is better to render 1 too many pixels (sometimes), rather than rendering too few.

Change in memory:
Flash: +18 bytes
SRAM: 0 bytes

gudnimg committed:
PFW-1504 rename length_step_mm to mm_per_pixel

Its more descriptive

gudnimg committed:
PFW-1504 More correct solution to the pixel count issue

The fact that the relationship between
machine position and pixel index is not linear
means we cannot simply rely on comparing
the previous position to the next machine position. i.e derivative of E_AXIS
Because around the max amplitude of the triangle wave
the slope will suddenly change sign and will create a deadzone
which has width 2*mm_per_pixel.
For MMU2S this is ~10mm (or two pixels).

Instead we should split the moves.
And only plan the 2nd move once we're
at the top of the triangle wave. That way we don't really care about the position delta.

Now we just calculate the current y(x)
position relative to current position and divide by mm_per_pixel.
If there is a delta measured with unit 'pixel' then that means
 its time to render the next pixel.

 This solution seems to work well so far on my end.

 Change in memory:
 Flash: +2 bytes
 SRAM: 0 bytes

gudnimg committed:
PFW-1504 Minor optimisation

Change in memory:
Flash: -4 bytes
SRAM: 0 bytes
